src/main/java/org/mariadb/jdbc/internal/failover/AbstractMastersListener.java

```diff
@@ -96,6 +99,24 @@ protected AbstractMastersListener(UrlParser urlParser) {
         this.urlParser = urlParser;
         this.masterHostFail.set(true);
     }
+    
+    protected void shutdownScheduler() {
+        schedulerProvider.shutdownScheduler(scheduler);
+        FailLoop runningLoop = stopFailover();
+        if (runningLoop != null) {
+            runningLoop.blockTillTerminated();
+        }
+    }
+    
+    @Override
+    protected void finalize() throws Throwable {
```

Comment:
I forgot to mention I modified the checkstyle to allow this finalizer.  This seems like a pretty important check to me, since it could result in leaked threads if not invoked (leaked threads were also possible before if the scheduler was not shutdown, but not being verified)